WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadStarter.kt

```diff
@@ -141,40 +133,41 @@ class UploadStarter @Inject constructor(
 
     /**
      * This is meant to be used by [checkConnectionAndUpload] only.
+     *
+     * The method needs to be synchronized from the following reasons. When the app comes to foreground both
+     * `queueUploadFromAllSites` and `queueUploadFromSite` are invoked. The problem is that they can run in parallel
+     * and `uploadServiceFacade.isPostUploadingOrQueued(it)` might return out-of-date result and a same post is added
+     * twice.
      */
+    @Synchronized
```

Comment:
Just noting an observation that I didn't notice before. 

The intent of `launch`-ing in here: 

https://github.com/wordpress-mobile/WordPress-Android/blob/aed0c79108ba08424af38b3eb28c2772a98bdc40/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadStarter.kt#L135-L140

... was so we could run `SELECT` queries in parallel, and then queue all the posts in parallel. But because of the concurrency issues in `isPostUploadingOrQueued()`, we wrapped this in a Mutex which essentially negated the `launch(ioDispatcher)`. 🤷‍♂ 

I think I overengineered `UploadStarter` in the first place anyway so it doesn't bother me that much. 😄 We _could_ just simplify `UploadStarter` in the future or make `UploadService` able to handle concurrency. It's a shame that we can't be concurrent if phones have multiple processors now.

Reply:
Good catch:D! I agree we could use just one launch and not use a launch in a foreach loop. However, the UploadService uploads the items serially + the data bandwidth is limited, so I'm not sure it has any real impact on the UX.